gateway/src/main/java/org/georchestra/gateway/security/oauth2/CustomOAuth2ClientPropertiesRegistrationAdapter.java

```diff
@@ -0,0 +1,111 @@
+import org.springframework.boot.autoconfigure.security.oauth2.client.OAuth2ClientProperties.Provider;
+import org.springframework.boot.context.properties.PropertyMapper;
+import org.springframework.boot.convert.ApplicationConversionService;
+import org.springframework.core.convert.ConversionException;
+import org.springframework.security.config.oauth2.client.CommonOAuth2Provider;
+import org.springframework.security.oauth2.client.registration.ClientRegistration;
+import org.springframework.security.oauth2.client.registration.ClientRegistration.Builder;
+import org.springframework.security.oauth2.client.registration.ClientRegistrations;
+import org.springframework.security.oauth2.core.AuthenticationMethod;
+import org.springframework.security.oauth2.core.AuthorizationGrantType;
+import org.springframework.security.oauth2.core.ClientAuthenticationMethod;
+import org.springframework.util.StringUtils;
+
+public final class CustomOAuth2ClientPropertiesRegistrationAdapter {
+
+    private CustomOAuth2ClientPropertiesRegistrationAdapter() {
+    }
+
+    public static Map<String, ClientRegistration> getClientRegistrations(OAuth2ClientProperties properties) {
+        Map<String, ClientRegistration> clientRegistrations = new HashMap<>();
+        properties.getRegistration().forEach((key, value) -> clientRegistrations.put(key,
+                getClientRegistration(key, value, properties.getProvider())));
+        return clientRegistrations;
+    }
+
+    private static ClientRegistration getClientRegistration(String registrationId,
+            OAuth2ClientProperties.Registration properties, Map<String, Provider> providers) {
+        Builder builder = getBuilderFromIssuerIfPossible(registrationId, properties.getProvider(), providers);
+        if (builder == null) {
+            builder = getBuilder(registrationId, properties.getProvider(), providers);
+        }
+        PropertyMapper map = PropertyMapper.get().alwaysApplyingWhenNonNull();
+        map.from(properties::getClientId).to(builder::clientId);
+        map.from(properties::getClientSecret).to(builder::clientSecret);
+        map.from(properties::getClientAuthenticationMethod).as(ClientAuthenticationMethod::new)
+                .to(builder::clientAuthenticationMethod);
+        map.from(properties::getAuthorizationGrantType).as(AuthorizationGrantType::new)
+                .to(builder::authorizationGrantType);
+        map.from(properties::getRedirectUri).to(builder::redirectUri);
+        map.from(properties::getScope).as(StringUtils::toStringArray).to(builder::scope);
+        map.from(properties::getClientName).to(builder::clientName);
+        if (registrationId.equals("franceconnect")) {
+            Map<String, Object> configurationMetadata = Collections.singletonMap("end_session_endpoint",
+                    "https://fcp.integ01.dev-franceconnect.fr");
+            map.from(configurationMetadata).to(builder::providerConfigurationMetadata);
+        }
```

Comment:
I think we have no choice than having hardcoded things for franceconnect unfortunately, but won't this be an issue when switching to "franceconnect-production" ? `integ01.dev` sounds quite dedicated to development purposes

Reply:
Discussed during the daily, another approach needs to be tested by @marwanehcine